### ISSUE TITLE
Lambda SQS Batch Error Handler Middleware

### ISSUE THREAD

**Issue by brettstack** [middyjs/middy#467]
Hi team, I created a middleware https://github.com/brettstack/civ6-play-by-cloud-turn-notifier/tree/master/packages/lambda-batch-sqs-error-cleaner-middleware to capture the pattern discussed here https://twitter.com/AWSbrett/status/1213542155873947648.

With the Lambda SQS Trigger, either all messages in the batch get deleted from the queue (when the Lambda resolves successfully) or none get deleted and remain on the queue to be processed again (when the Lambda throws an error).

In the case where at least 1 message wasn't successfully processed, this middleware handles deleting the successfully processed messages for you, then throws an error to return the remaining messages to the queue. If all messages are processed successfully, the middleware doesn't delete the messages itself, but simply resolves and lets the Lambda service handle deleting the messages (outside of your function invocation).

The middleware determines successfully processed messages and failures based on the handler returning a `Promise.allSettled` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled result.

```js
const recordPromises = Records.map(async (record, index) => { /* Custom message processing logic */ })
const settledRecords = await Promise.allSettled(recordPromises)
return settledRecords
```

I'd appreciate any feedback you have. Let me know if you're interested in adding this to the core middleware in this repo.

**Comment by lmammino** [comment]
Hello @brettstack, this looks great! Thanks for taking the time to formalize this here.

I am supportive for having this integrate into the core set of middlewares (for middy 1.0.0) , but i'd be curious to know how common this use case can be. If you believe that there will be a significative user base for it, than it makes sense to do extra due diligence to have this package always in sync and tested together with the rest of the core middlewares.

If not, it would still be good to have this well structured, documented and available on NPM so that we can add it to the list of third-party middlewares.

@willfarrell, what do you think?

**Comment by brettstack** [comment]
This is the only data I have on how common this: https://stackoverflow.com/questions/54596129/processing-aws-lambda-messages-in-batches/54611899#54611899, which isn't a very popular SO question.

Partially failed batches is something everyone using SQS Batch processing will/should have to deal with. However, since users are already advised to check if a particular message has already been processed  (since SQS is at-least-once delivery), this issue is sometimes masked (especially if the developer is manually deleting the message at that point).

**Comment by willfarrell** [comment]
I'm all for this as part of the core offering for v1. Using SQS is a fundamental piece of building scalable serverless cloud applications.

This reminds me of #173 and our need to have complete best practice examples and should be added to the list w/ a diagram.

**Comment by brettstack** [comment]
@willfarrell @lmammino I've opened the PR here #471. 

On this:

> then it makes sense to do extra due diligence to have this package always in sync and tested together with the rest of the core middlewares

Do you have examples of other core middleware that test their interaction with others?

**Comment by lmammino** [comment]
Closed by #471
